Add PM usage counter to "device list"

fabiobaltieri opened the pull request:
Useful for debugging, example (with https://github.com/zephyrproject-rtos/zephyr/pull/71788 as well):

```console
uart:~$ device list
devices:
- buttons (suspended, usage=0)
uart:~$ pm runtime-get buttons 
uart:~$ pm runtime-get buttons
uart:~$ device list
devices:
- buttons (active, usage=2)
uart:~$ pm runtime-put buttons
uart:~$ pm runtime-put buttons
uart:~$ device list
devices:
- buttons (suspended, usage=0)
```

fabiobaltieri committed:
pm: device_runtime: add an access function to get the usage counter

Add a pm_device_usage_get() to get the current usage counter.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
shell: device_service: show the usage conter when pm runtime is enabled

Add an extra "usage=%d" to the device status to show the current usage
counter when PM runtime is enabled.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

bjarki-trackunit commented:
Also big fan :)

fabiobaltieri commented:
> I totally want this called out in the docs too :D This is awesome sauce!

should have seen that coming :-) let's wrap https://github.com/zephyrproject-rtos/zephyr/pull/71788 and then I can add a paragraph under "Device PM testing" that that PR introduce, moving this to draft in the meantime.

gmarull commented on include/zephyr/pm/device_runtime.h:
```diff
@@ -158,6 +159,25 @@ int pm_device_runtime_put_async(const struct device *dev, k_timeout_t delay);
+ *
+ * @param dev Device instance.
+ *
+ * @retval the current usage counter
```
```suggestion
 * @returns the current usage counter
```

gmarull commented on subsys/shell/modules/device_service.c:
```diff
@@ -79,7 +79,17 @@ static int cmd_device_list(const struct shell *sh,
 		}
 
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+		if (dev->pm != NULL) {
+			uint32_t usage = pm_device_usage_get(dev);
```
you should avoid accessing dev->pm, just provide a dummy impl for pm_device_usage_get in case PM is not built-in. Function could return int, <0 for error (-ENOSYS), >=0 for usage counter. It will also avoid ifdeffery in clients.

fabiobaltieri replied:
wow, what was I thinking... there were other issues as well, namely returning a counter for device with normal pm, non-runtime, what's the opposite of "runtime" in this context? whatever, fixed, should handle all cases now, example:

```console
uart:~$ device list 
devices:
- clock@40000000 (READY)
- gpio@50000000 (READY)
- random@4000d000 (READY)
- entropy_bt_hci (READY)
- uart@40002000 (active)
- flash-controller@4001e000 (READY)
- i2c@40003000 (active)
- buttons (suspended, usage=0)
- leds (READY)
```

fabiobaltieri committed:
doc: services: pm: document the "device list" command

Document the "device list" command in the context of device power
management.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Alright rebased and added a paragraph in the documentation.

gmarull commented on include/zephyr/pm/device_runtime.h:
```diff
@@ -10,6 +10,7 @@
 
 #include <zephyr/device.h>
 #include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
```
why is this needed?

fabiobaltieri replied:
good catch, it's a leftover for getting `struct pm_device` when this was an inline, but then I moved it to the normal file so this can go now

fabiobaltieri commented on include/zephyr/pm/device_runtime.h:
```diff
@@ -158,6 +159,17 @@ int pm_device_runtime_put_async(const struct device *dev, k_timeout_t delay);
+ * @retval -EINVAL If the device is not using runtime PM.
+ * @retval -ENOSYS If the runtime PM is not enabled at all.
+ */
+int pm_device_usage_get(const struct device *dev);
```
yeah makes sense

fabiobaltieri replied:
...or should it be just `pm_device_runtime_usage`? since there's already a `get` that does something else

fabiobaltieri replied:
yeah let me drop the `get` bit, it's just confusing

fabiobaltieri commented on doc/services/pm/device.rst:
```diff
@@ -249,6 +249,22 @@ option and issue a ``pm`` command on a device from the shell, for example:
         devices:
         - buttons (suspended)
 
+To print the power management state of a device, enabled the
```
fixed

fabiobaltieri commented:
- renamed the function to `pm_device_runtime_usage` (and did the other suggestions)

gmarull commented on include/zephyr/pm/device_runtime.h:
```diff
@@ -158,6 +158,17 @@ int pm_device_runtime_put_async(const struct device *dev, k_timeout_t delay);
+ * @param dev Device instance.
+ *
+ * @returns the current usage counter.
+ * @retval -EINVAL If the device is not using runtime PM.
```
```suggestion
 * @retval -ENOTSUP If the device is not using runtime PM.
```

fabiobaltieri committed:
tests: shell: add tests for "device list"

Add tests for the "device list" output with the various configuration of
device power management.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
Added tests for this one as well. Definitely not because I misread what PR @henrikbrixandersen asked me to add tests for.

fabiobaltieri commented:
@ceolin can you do another review please? Same as what you've seen already but with tests.

ceolin commented:
> @ceolin can you do another review please? Same as what you've seen already but with tests.

Thank you !
